File: src/modules/LightSwitch/Tests/LightSwitch.Cli.UnitTests/CliProtocolTests.cs
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

using System;
using LightSwitch.Cli.Protocol;
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace LightSwitch.Cli.UnitTests;

[TestClass]
public sealed class CliProtocolTests
{
    private static readonly byte[] ExpectedEncodedLine = { 0x41, 0x00, 0x0A, 0x00 };

    [TestMethod]
    [DataRow("null")]
    [DataRow("not-json")]
    [DataRow("{}")]
    [DataRow("{\"version\":1,\"state\":{}}")]
    [DataRow("{\"version\":1,\"success\":true}")]
    [DataRow("{\"version\":1,\"success\":false}")]
    [DataRow("{\"version\":2,\"success\":false,\"error\":{\"code\":\"EXECUTION_FAILED\",\"message\":\"Failed.\"}}")]
    [DataRow("{\"success\":false,\"error\":{\"code\":\"EXECUTION_FAILED\",\"message\":\"Failed.\"}}")]
    [DataRow("{\"version\":1,\"success\":false,\"error\":{\"code\":\"\",\"message\":\"Failed.\"}}")]
    [DataRow("{\"version\":1,\"success\":false,\"error\":{\"code\":\"EXECUTION_FAILED\",\"message\":null}}")]
    [DataRow("{\"version\":1,\"success\":false,\"error\":{\"code\":\"EXECUTION_FAILED\"}}")]
    [DataRow("{\"version\":1,\"success\":false,\"error\":{\"message\":\"Failed.\"}}")]
    public void InvalidEnvelopesAreProtocolErrors(string json)
    {
        var error = Assert.ThrowsException<CliException>(() => CliProtocol.ParseResponse(json));
        Assert.AreEqual("PROTOCOL_ERROR", error.Code);
    }

    [TestMethod]
    [DataRow("systemTheme", "null")]
    [DataRow("systemTheme", "\"sepia\"")]
    [DataRow("appsTheme", "\"Dark\"")]
    [DataRow("scheduleMode", "\"automatic\"")]
    [DataRow("changeSystem", "\"true\"")]
    public void InvalidStateMembersAreRejected(string property, string replacement)
    {
        string original = property switch
        {
            "systemTheme" => "\"light\"",
            "appsTheme" => "\"dark\"",
            "scheduleMode" => "\"FixedHours\"",
            _ => "true",
        };
        string json = CliApplicationTests.SuccessfulResponse.Replace("\"" + property + "\":" + original, "\"" + property + "\":" + replacement, StringComparison.Ordinal);

        var error = Assert.ThrowsException<CliException>(() => CliProtocol.ParseResponse(json));
        Assert.AreEqual("PROTOCOL_ERROR", error.Code);
    }

    [TestMethod]
    public void MissingFalseValuedStateMembersAreNotSilentlyDefaulted()
    {
        string json = CliApplicationTests.SuccessfulResponse.Replace("\"changeApps\":false,", string.Empty, StringComparison.Ordinal);
        Assert.ThrowsException<CliException>(() => CliProtocol.ParseResponse(json));
    }

    [TestMethod]
    public void ContradictorySuccessAndErrorAreRejected()
    {
        string json = CliApplicationTests.SuccessfulResponse.Replace("\"success\":true,", "\"success\":true,\"error\":{\"code\":\"EXECUTION_FAILED\",\"message\":\"Failed.\"},", StringComparison.Ordinal);
        Assert.ThrowsException<CliException>(() => CliProtocol.ParseResponse(json));
    }

    [TestMethod]
    public void UnknownThemeIsAValidState()
    {
        string json = CliApplicationTests.SuccessfulResponse.Replace("\"systemTheme\":\"light\"", "\"systemTheme\":\"unknown\"", StringComparison.Ordinal);
        Assert.AreEqual("unknown", CliProtocol.ParseResponse(json).State!.SystemTheme);
    }

    [TestMethod]
    public void ErrorCanIncludeTheResultingState()
    {
        string json = CliApplicationTests.SuccessfulResponse.Replace("\"success\":true,", "\"success\":false,\"error\":{\"code\":\"EXECUTION_FAILED\",\"message\":\"Partially applied.\"},", StringComparison.Ordinal);
        var response = CliProtocol.ParseResponse(json);
        Assert.IsFalse(response.Success);
        Assert.IsNotNull(response.State);
        Assert.AreEqual("EXECUTION_FAILED", response.Error!.Code);
    }

    [TestMethod]
    public void OversizedEnvelopeIsRejectedBeforeDeserialization()
    {
        Assert.ThrowsException<CliException>(() => CliProtocol.ParseResponse(new string(' ', CliProtocol.MaxMessageChars + 1)));
    }

    [TestMethod]
    public void PipeEncodingHasNoBomAndUsesLittleEndianUtf16()
    {
        Assert.AreEqual(0, CliProtocol.Encoding.GetPreamble().Length);
        CollectionAssert.AreEqual(ExpectedEncodedLine, CliProtocol.Encoding.GetBytes("A\n"));
    }
}

File: src/modules/LightSwitch/Tests/LightSwitchService.UnitTests/CliProtocolTests.cpp
// Copyright (c) Microsoft Corporation
// The Microsoft Corporation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#include <CppUnitTest.h>
#include "CliProtocol.h"
#include "TestSupport.h"

#include <array>
#include <utility>

using namespace Microsoft::VisualStudio::CppUnitTestFramework;
using namespace light_switch_cli;
using namespace winrt::Windows::Data::Json;

namespace LightSwitchServiceUnitTests
{
    TEST_CLASS (CliProtocolTests)
    {
    public:
        TEST_METHOD (AcceptsSupportedCommandsAndExplicitModes)
        {
            Apartment apartment;
            const std::array<std::pair<std::wstring_view, Command>, 6> commands{
                { { L"status", Command::Status }, { L"light", Command::Light }, { L"dark", Command::Dark }, { L"toggle", Command::Toggle }, { L"schedule-enable", Command::ScheduleEnable }, { L"schedule-disable", Command::ScheduleDisable } }
            };
            for (const auto& command : commands)
            {
                const auto request = ParseRequest(L"{\"version\":1,\"command\":\"" + std::wstring(command.first) + L"\"}");
                Assert::IsTrue(command.second == request.command);
                Assert::IsFalse(request.mode.has_value());
            }
            for (const auto mode : { L"FixedHours", L"SunsetToSunrise", L"FollowNightLight" })
            {
                const auto request = ParseRequest(L"{\"version\":1,\"command\":\"schedule-enable\",\"mode\":\"" + std::wstring(mode) + L"\"}");
                Assert::IsTrue(request.command == Command::ScheduleEnable);
                Assert::AreEqual(std::wstring(mode), request.mode.value());
            }
        }

        TEST_METHOD (RejectsUnknownOrIncompatibleArguments)
        {
            Apartment apartment;
            for (const auto message : {
                     L"{\"version\":1,\"command\":\"unknown\"}",
                     L"{\"version\":1,\"command\":true}",
                     L"{\"version\":1}",
                     L"{\"version\":1,\"command\":\"status\",\"mode\":\"FixedHours\"}",
                     L"{\"version\":1,\"command\":\"schedule-disable\",\"mode\":\"FixedHours\"}",
                     L"{\"version\":1,\"command\":\"schedule-enable\",\"mode\":null}",
                     L"{\"version\":1,\"command\":\"schedule-enable\",\"mode\":\"Off\"}",
                     L"{\"version\":1,\"command\":\"schedule-enable\",\"mode\":\"fixedhours\"}",
                     L"{\"version\":1,\"command\":\"status\",\"extra\":0}" })
            {
                AssertRequestError(message, L"INVALID_ARGUMENT");
            }
        }

        TEST_METHOD (RejectsMalformedAndUnsupportedProtocol)
        {
            Apartment apartment;
            for (const auto message : {
                     L"", L"not json", L"[]", L"null", L"{", L"{\"command\":\"status\"}", L"{\"version\":2,\"command\":\"status\"}", L"{\"version\":1.5,\"command\":\"status\"}", L"{\"version\":\"1\",\"command\":\"status\"}", L"{\"version\":1,\"command\":\"status\"} {}", L"{\"version\":1,\"command\":\"status\",\"command\":\"toggle\"}", L"{\"version\":1,\"command\":\"status\",\"\\u0063ommand\":\"toggle\"}" })
            {
                AssertRequestError(message, L"PROTOCOL_ERROR");
            }
        }

        TEST_METHOD (EnforcesBomlessUtf16AndMessageLength)
        {
            Apartment apartment;
            const std::wstring valid = L"{\"version\":1,\"command\":\"status\"}";
            AssertRequestError(std::wstring(1, L'\xfeff') + valid, L"PROTOCOL_ERROR");
            AssertRequestError(valid + std::wstring(1, L'\0'), L"PROTOCOL_ERROR");
            AssertRequestError(valid + std::wstring(1, L'\xd800'), L"PROTOCOL_ERROR");
            AssertRequestError(valid + std::wstring(1, L'\xdc00'), L"PROTOCOL_ERROR");
            AssertRequestError(std::wstring(MaxMessageCharacters + 1, L' '), L"PROTOCOL_ERROR");
            AssertRequestError(std::wstring(16000, L'[') + std::wstring(16000, L']'), L"PROTOCOL_ERROR");
            const auto atLimit = std::wstring(MaxMessageCharacters - valid.size(), L' ') + valid;
            Assert::IsTrue(ParseRequest(atLimit).command == Command::Status);
        }

        TEST_METHOD (ResponseEnvelopesRetainStateAndStructuredError)
        {
            Apartment apartment;
            const auto state = TestState();
            const auto success = MakeSuccess(state);
            Assert::AreEqual(1.0, success.GetNamedNumber(L"version"));
            Assert::IsTrue(success.GetNamedBoolean(L"success"));
            Assert::IsFalse(success.HasKey(L"error"));
            Assert::AreEqual(std::wstring(L"unknown"), std::wstring(success.GetNamedObject(L"state").GetNamedString(L"appsTheme")));

            const auto failure = MakeError(L"EXECUTION_FAILED", L"A test failure", state);
            Assert::IsFalse(failure.GetNamedBoolean(L"success"));
            Assert::IsTrue(failure.HasKey(L"state"));
            Assert::AreEqual(std::wstring(L"EXECUTION_FAILED"), std::wstring(failure.GetNamedObject(L"error").GetNamedString(L"code")));
            Assert::IsFalse(MakeError(L"INVALID_ARGUMENT", L"Invalid request").HasKey(L"state"));
        }

    private:
        static void AssertRequestError(std::wstring_view message, std::wstring_view expectedCode)
        {
            try
            {
                ParseRequest(message);
                Assert::Fail(L"The request should have been rejected.");
            }
            catch (const RequestError& error)
            {
                Assert::AreEqual(std::wstring(expectedCode), error.Code());
                Assert::IsFalse(error.Message().empty());
            }
        }
    };
}
